app/services/publish.js

```diff
@@ -0,0 +1,106 @@
+import Service, { inject as service } from '@ember/service';
+import { task, all } from 'ember-concurrency';
+import { tracked } from '@glimmer/tracking';
+
+export class Extract {
+  @tracked treatmentId;
+  @tracked position;
+  @tracked document;
+  @tracked errors;
+
+  constructor(treatmentId, position, document, errors) {
+    this.treatmentId = treatmentId;
+    this.position = position;
+    this.document = document;
+    this.errors = errors;
+  }
+}
+
+export default class PublishService extends Service {
+  @tracked treatmentExtractsMap;
+  @service store;
+
+  constructor(...args) {
+    super(...args);
+    this.treatmentExtractsMap = new Map();
+  }
+
+  get treatmentExtracts() {
+    return [...this.treatmentExtractsMap.values()].sortBy('position');
+  }
+
+  get isIdle() {
+    return this.loadExtractsTask.isIdle;
+  }
+
+  get loadExtractsTask() {
+    return this._loadExtractsTask.unlinked();
+  }
+
+  /**
+   * Combine saved extracts with newly created ones and expose them as one map keyed by the
+   * id of the treatment
+   * @param meetingId
+   * @return {Generator<*, void, *>}
+   */
+  @task
+  *_loadExtractsTask(meetingId) {
+    const [newExtracts, meeting, treatments, versionedTreatments] = yield all([
+      fetch(`/prepublish/behandelingen/${meetingId}`).then((res) => res.json()),
+      this.store.findRecord('zitting', meetingId),
+      this.store.query('behandeling-van-agendapunt', {
+        'filter[onderwerp][zitting][:id:]': meetingId,
+        include: 'onderwerp',
+      }),
+      this.store.query('versioned-behandeling', {
+        'filter[zitting][:id:]': meetingId,
+        include: 'behandeling.onderwerp,signed-resources,published-resource',
+      }),
+    ]);
+
+    // map all treatments to their ids for later retrieval
+    const treatmentMap = new Map();
+    treatments.forEach((treatment) =>
+      treatmentMap.set(treatment.id, treatment)
+    );
+
+    // map all existing extracts to their behandeling id
+    const extractMap = new Map();
+    versionedTreatments.forEach((versionedTreatment) => {
+      extractMap.set(
+        versionedTreatment.behandeling.get('id'),
+        new Extract(
+          versionedTreatment.behandeling.get('id'),
```

Comment:
I have a feeling this code is not sound. It works because behandeling is included, but I don't know if it is guaranteed to work, since technically this is still a relationship and should thus be treated as being async.

I think the proper way to solve this would be to yield the behandeling, and rely on the include to avoid the unnecessary network request. However, since nothing in ember data is iterable (that would just be sensible, and of course we don't ever want anything sensible) this is really annoying to do (you can't yield from a forEach).

@nvdk @Windvis any suggestions?

Reply:
Always resolving the async hasMany relationship before accessing its data does seem to be the best practice. I've only recently learned of this myself since there is an [open RFC](https://github.com/runspired/rfcs/blob/ember-data/deprecate-methods-on-promise-many-array/text/0745-ember-data-deprecate-methods-on-promise-many-array.md) to deprecate the direct data access. The RFC only mentions hasMany relationships though.. not sure if that's different from belongsTo (and I don't really know how to find that out either besides asking in the Discord). I'm still far away from being an Ember Data expert, sorry :smile:.

> since nothing in ember data is iterable

Do you mean you can't do `for (const treatment of versionedTreatments)` because `versionedTreatments` is not iterable? I _think_ you can just call [`toArray()`](https://api.emberjs.com/ember/3.26/classes/Ember.NativeArray/methods/toArray?anchor=toArray) on to it first?

Anyway, I still think the real solution is just converting everything to sync relationships since that will make things a lot more straightforward. I really hoped they would have made that the default in 4.0 since all the core members seem to agree on that at least. I heard you like going on refactoring sprees? :grimacing:

Reply:
I also think it's ok to leave it like this since it works. We can always deal with the deprecations later if needed.

If you want to be really sure you could always post in the #ember-data channel in the Discord. Runspired is very active there and will certainly have the answer for you.

Reply:
Yeah I think at least in GN going all sync would have a lot of value
I get what the async relations are going for but the syntax looks way too much like normal js on the one hand, and is annoying on the other (like in this case, you always have to save every relationship to a variable, you cant use the already existing model object)

Reply:
gonna leave it, too close to release to muck about